torch/lib/THCUNN/LookupTable.cu

```diff
@@ -193,5 +193,39 @@ struct multiply_s
   }
 };
 
+template <typename Dtype, typename Acctype>
+struct renorm_functor
+{
+  thrust::device_ptr<Dtype> weight_ptr;
+  int64_t stride;
+  Dtype maxNorm;
+  Dtype normType;
+  thrust::plus<Acctype> binary_plus;
+  pow_v<Dtype, Acctype> unary_pow;
+
+  renorm_functor(
+      thrust::device_ptr<Dtype> w, int64_t s, Dtype m, Dtype n, 
+      thrust::plus<Acctype> bp, pow_v<Dtype, Acctype> up
+  ) : weight_ptr(w), stride(s), maxNorm(m), normType(n), 
+      binary_plus(bp), unary_pow(up) {}
+
+  #pragma hd_warning_disable
+  __host__ __device__
+  void operator()(const THCIndex_t idx)
+  {
+    THCIndex_t k  = idx - TH_INDEX_BASE;
+    thrust::device_ptr<Dtype> row_ptr = weight_ptr + k * stride;
+    Acctype norm = thrust::transform_reduce(thrust::device, row_ptr, row_ptr + stride,
```

Comment:
Unfortunately, I don't think this is the right solution.

In this case, each individual CUDA thread would be performing a single reduction on an entire vector, which means that all of the memory loads by that thread will be non-coalesced since the reads are strided across the warp by a rather large stride, so the kernel would be performing up to 1/32x slower than it needs to perform. 

The proper solution to parallelize the outer loop below unfortunately would be to write a custom kernel here that calculates the norm on a per-block basis, then goes back and transforms the data. The outer dimension (numel) is handled over the block index. Norm calculation would be a block-wise reduction. All threads in the block would load from the vector in a coalesced manner, and would then go back and update the data.

The current code has a rather large kernel launch overhead, but does have the coalesced accesses. You have eliminated the kernel launch overhead but have made the memory accesses worse. Just as the old version would perform rather well for a small batch size but massive vectors, your version would perform poorly if the vectors were really huge.

To "correctly" solve a problem like this might require several different specializations (large batch / small dim, small dim / large batch, both large), but giving each vector to a single block is a good middle ground, and should be ok for pretty much anything.